### PR TITLE
refactor(compaction): unify task cancellation

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Refactor compaction scheduler's large method `pick_and_assign`, to make test easier.

- Remove the loop which retries selecting, assigning and sending task.
- Extract task cancellation.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/meta/src/hummock/compaction_scheduler.rs [hunk 1]
```diff
@@ -137,127 +148,344 @@ where
...
     async fn pick_and_assign(
         &self,
         compaction_group: CompactionGroupId,
         request_channel: Arc<CompactionRequestChannel>,
-    ) -> bool {
+    ) -> ScheduleStatus {
+        let schedule_status = self
+            .pick_and_assign_impl(compaction_group, request_channel)
+            .await;
+        let cancel_task = match &schedule_status {
```
Comment (Li0k) [inline src/meta/src/hummock/compaction_scheduler.rs:162]:
```suggestion
        let cancel_task = match &schedule_status {
            ScheduleStatus::NoAvailableCompactor(task) |  ScheduleStatus::AssignFailure(task) | ScheduleStatus::SendFailure(task) => Some(task.clone()),
            _ => None,
        }
```

File: src/meta/src/hummock/compaction_scheduler.rs [hunk 2]
```diff
@@ -137,127 +148,344 @@ where
...
+            ScheduleStatus::PickFailure => None,
+        };
+        if let Some(mut compact_task) = cancel_task {
+            // Try to cancel task immediately.
+            if self
+                .hummock_manager
+                .cancel_compact_task(&mut compact_task)
+                .await
+                .is_err()
+            {
```
Comment (Li0k) [inline src/meta/src/hummock/compaction_scheduler.rs:177]: we cancel task asynchronously, how can we track it if `cancel_compact_task` fail? Maybe need some log / metrics here ?

File: src/meta/src/hummock/manager/compaction.rs [hunk 3]
```diff
@@ -122,4 +122,18 @@ where
             .expect("compaction group exists")
             .compaction_config()
     }
+
+    #[named]
+    pub async fn list_all_tasks(&self) -> Vec<HummockCompactionTaskId> {
```
Comment (Li0k) [inline src/meta/src/hummock/manager/compaction.rs:127]:
```suggestion
    pub async fn list_all_task_ids(&self) -> Vec<HummockCompactionTaskId> {
```

File: src/meta/src/hummock/compaction_scheduler.rs (hunk 1 above)
Comment (zwang28) [reply src/meta/src/hummock/compaction_scheduler.rs:162]: Fixed

File: src/meta/src/hummock/manager/compaction.rs (hunk 3 above)
Comment (zwang28) [reply src/meta/src/hummock/manager/compaction.rs]: Fixed.

File: src/meta/src/hummock/compaction_scheduler.rs (hunk 2 above)
Comment (zwang28) [reply src/meta/src/hummock/compaction_scheduler.rs:175]: Fixed.

Comment (zwang28) [reply src/meta/src/hummock/compaction_scheduler.rs:175]: We log whenever cancel fails, either here or in the async worker.

**Review by Li0k — APPROVED** [review]
LGTM

**Pull request merged**